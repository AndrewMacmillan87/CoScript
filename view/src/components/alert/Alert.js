import React from "react";

// Redux
import { withRouter } from "react-router-dom";
import { connect } from "react-redux";

import PropTypes from "prop-types";

import "./alert.css";

export const Alert = withRouter(({ alerts, userProfile, projectData, ...props }) => {
    return (
        alerts !== null &&
        alerts.length > 0 &&
        alerts.map((alert) => {
            switch (props.location.pathname) {
                case "/login":
                    return (
                        <div key={alert.id} className={`login-alert alert-${alert.alertType}`}>
                            {alert.msg}
                        </div>
                    );
                case "/account":
                    return (
                        <div key={alert.id} className={`account-alert alert-${alert.alertType}`}>
                            {alert.msg}
                        </div>
                    );
                case "/update-credentials":
                    return (
                        <div key={alert.id} className={`update-alert alert-${alert.alertType}`}>
                            {alert.msg}
                        </div>
                    );
                case `/landing${projectData ? projectData.url : ""}/contact`:
                    return (
                        <div key={alert.id} className={`contact-alert alert-${alert.alertType}`}>
                            {alert.msg}
                        </div>
                    );
                case "/profile":
                    return (
                        <div key={alert.id} className={`delete-alert alert-${alert.alertType}`}>
                            {alert.msg}
                        </div>
                    );
                case "/messages":
                case "/public-profile":
                    return (
                        <div key={alert.id} className={`message-alert alert-${alert.alertType}`}>
                            {alert.msg}
                        </div>
                    );
                case `/profile/user/${userProfile ? userProfile.id : ""}`:
                    return (
                        <div key={alert.id} className={`request-alert alert-${alert.alertType}`}>
                            {alert.msg}
                        </div>
                    );
                default:
                    return (
                        <div key={alert.id} className={`alert alert-${alert.alertType}`}>
                            {alert.msg}
                        </div>
                    );
            }
        })
    );
});

Alert.propTypes = {
    alerts: PropTypes.array.isRequired,
    userProfile: PropTypes.object,
    projectData: PropTypes.object,
};

// The state here comes from the rootReducer - reducers/index.js
const setProps = (state) => ({
    alerts: state.alertpopup,
    userProfile: state.auth.userProfile,
    projectData: state.documents.projectData,
});

export default connect(setProps)(Alert);
