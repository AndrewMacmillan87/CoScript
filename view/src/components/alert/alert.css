.alert,
.login-alert,
.account-alert,
.update-alert,
.contact-alert,
.message-alert,
.request-alert,
.delete-alert {
    text-align: center;
    font-size: 0.9rem;
    padding: 0.6rem;
    margin: 5px auto;
    opacity: 0.9;
    background-color: var(--background-color-mid, #444);
    color: #fff;
}

.alert {
    width: 80%;
}

.login-alert {
    margin-left: 0 !important;
    width: 77%;
}

.account-alert {
    margin: 15px auto 5px auto;
    padding: 0.6rem 0 !important;
    width: 100%;
}

.update-alert {
    width: 100%;
    margin-top: 20px;
    padding: 0.6rem 0 !important;
}

.contact-alert {
    width: 90%;
    margin-top: 20px;
    margin-left: 0;
    padding: 10px 10px 10px 5px !important;
}

.message-alert {
    width: 100vw - 300px;
    margin: auto;
    padding: 10px 10px 10px 5px !important;
}

.request-alert {
    position: fixed;
    top: 70px;
    width: 100vw;
    margin: auto;
    padding: 10px 10px 10px 5px !important;
}

.delete-alert {
    position: fixed;
    top: 50px;
    right: 50px;
    font-size: 1.3rem;
    width: 30vw;
    opacity: 1 !important;
    margin: auto;
    padding: 20px 20px 20px 15px !important;
}

.alert-danger {
    background-color: var(--danger, #fb1515);
    color: #fff;
}

.alert-success {
    background-color: var(--button-green, #4af030);
    color: #fff;
}
