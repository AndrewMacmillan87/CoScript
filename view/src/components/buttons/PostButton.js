import React from "react";
import PropTypes from "prop-types";

import "./buttons.css";

const PostButton = ({ onClick, classes, value }) => {
    return (
        <input
            onClick={onClick}
            className={classes}
            type='submit'
            value={value}
        />
    );
};

PostButton.defaultProps = {
    value: "Verify"
};

PostButton.propTypes = {
    onClick: PropTypes.func,
    classes: PropTypes.string,
    value: PropTypes.string
};

export default PostButton;
