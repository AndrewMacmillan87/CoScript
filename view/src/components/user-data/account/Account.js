import React from "react";
import { withRouter } from "react-router-dom";
import MobileNav from "../MobileNav";
import SidePanel from "../SidePanel";
import AccountInfo from "./AccountInfo";
import Password from "./Password";
import DeleteAccount from "./DeleteAccount";

// Redux
import { connect } from "react-redux";
import { displayAlert } from "../../../redux/actions/alertpopup";
import { sendPWEmail, verifyPWCode } from "../../../redux/auxiliary/sendEmail";
import { loadUser, deleteUser } from "../../../redux/actions/auth";
import { updateUserEmail } from "../../../redux/actions/account";

import PropTypes from "prop-types";

import "./account.css";

export class Account extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            email: "",
            password: "",
            confirmPassword: "",
            confirmCode: "",
            confirmDeleteCode: "",
            deleteCode: NaN,
            deleteCodeIncorrect: false,
            verifyFailedMsg: "",
            visible: false,
            authenticated: false,
            failed: false,
            displayEmailWindow: false,
            displayDeleteWindow: false,
            loading: false,
        };
        this.displaySpinner = this.displaySpinner.bind(this);
        this.removeSpinner = this.removeSpinner.bind(this);
        this.cancelEdit = this.cancelEdit.bind(this);
        this.hidePassword = this.hidePassword.bind(this);
        this.onClick = this.onClick.bind(this);
        this.onChange = this.onChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.sendChangeCode = this.sendChangeCode.bind(this);
        this.submitCode = this.submitCode.bind(this);
        this.onDelete = this.onDelete.bind(this);
        this.emailField = React.createRef();
        this.deleteField = React.createRef();
        this.verifyFailed = React.createRef();
    }

    /**
     * Sets the browser tab title and scrolls to the top
     * of the page when the component mounts
     */
    componentDidMount() {
        document.title = "CoScript - Account";
        window.scrollTo(0, 0);
    }

    /**
     * Checks the 'failed' state property to see if a
     * verification code failed to match and sets the
     * 'verifiedFailed' ref textContext to the value of
     * 'verifyFailedMsg' if it did
     */
    componentDidUpdate() {
        const { failed, verifyFailedMsg } = this.state;
        if (failed) {
            this.verifyFailed.current.textContent = verifyFailedMsg;
        }
    }

    /**
     * Sets the 'loading' state property to true to display
     * a loading spinner to the user while loading
     */
    displaySpinner() {
        this.setState({
            loading: true,
        });
    }

    /**
     * Sets the 'loading' state property to false to remove
     * a loading spinner from the UI
     */
    removeSpinner() {
        this.setState({
            loading: false,
        });
    }

    /**
     * Falsifies all state properties that are associated
     * with displaying UI elements to the user so they can
     * edit their account details
     */
    cancelEdit() {
        this.setState({
            displayEmailWindow: false,
            displayDeleteWindow: false,
            deleteCode: NaN,
            deleteCodeIncorrect: false,
        });
    }

    /**
     * Sets the 'visible' state property to false to hide the
     * password update verification code field. This also clears
     * the generated code out of local storage **THIS NEEDS TO
     * BE REVISED - CODES SHOULD BE STORED IN THE DATABASE**
     */
    hidePassword() {
        this.setState({
            visible: false,
        });
        if (localStorage.getItem("originalCode") !== null) {
            localStorage.removeItem("originalCode");
        }
    }

    /**
     * Displays an edit window to the user so they can
     * either update their email address or choose to
     * delete their account
     *
     * @param {Object} e - event object
     */
    onClick(e) {
        const { user } = this.props;
        switch (e.target.id) {
            case "edit-email-btn":
                this.setState({
                    displayEmailWindow: true,
                });
                setTimeout(() => {
                    this.emailField.current.value = user.email;
                }, 500);
                break;
            case "delete-account-btn":
                this.setState({
                    displayDeleteWindow: true,
                    deleteCode: Math.floor(Math.random() * 1000000) + 423523,
                });
                break;
            default:
                return;
        }
    }

    /**
     * Sets the <[e.target.name]> state property to the
     * value of the event target
     *
     * @param {Object} e - event object
     */
    onChange(e) {
        this.setState({
            [e.target.name]: e.target.value,
        });
    }

    /**
     * Updates the user's email address in the database
     *
     * @param {Object} e - event object
     */
    onSubmit(e) {
        e.preventDefault();
        this.setState({
            displayEmailWindow: false,
        });
        const { user, updateUserEmail } = this.props;
        const newAccountState = {
            ...user,
            email: this.state.email,
        };
        updateUserEmail(newAccountState);
    }

    /**
     * Sets the 'visible' state property to true to display
     * the verification code entry field to the user. Also
     * sends an email to the user with the code they have
     * to enter to change their password
     */
    async sendChangeCode() {
        this.setState({
            visible: true,
        });
        const { user } = this.props;
        const res = await sendPWEmail(user.email);
        setTimeout(() => {
            localStorage.setItem("originalCode", res.data.originalCode);
        }, 500);
    }

    /**
     * Checks the verification code the user entered
     * against the original code being stored in
     * local storage. If the codes match, the user is
     * give an auth token and redirected to a protected
     * page where they can update their password. If the
     * codes don't match, the user is shown an error
     * message
     *
     * @param {Object} e - event object
     */
    async submitCode(e) {
        e.preventDefault();
        this.displaySpinner();
        const { confirmCode } = this.state;
        const data = {
            originalCode: localStorage.getItem("originalCode"),
            confirmCode,
            forgottenDetails: false,
        };
        const res = await verifyPWCode(data);
        if (res.success) {
            await this.props.loadUser();
            localStorage.removeItem("originalCode");
            this.setState({
                failed: false,
                authenticated: true,
            });
            this.props.history.push("/update-credentials");
        } else {
            this.setState({
                failed: true,
                verifyFailedMsg: res.msg,
            });
            this.removeSpinner();
        }
    }

    /**
     * Checks to see if the code displayed to the user
     * when the window opens matches the code the user
     * enters into the input field. If it matches, the
     * user's account is deleted and they are redirected
     * to the homepage. If not, an error message is shown
     *
     * @param {Object} e - event object
     */
    onDelete(e) {
        e.preventDefault();
        const { user, history, deleteUser } = this.props;
        const { deleteCode, confirmDeleteCode } = this.state;
        if (deleteCode === Number(confirmDeleteCode)) {
            history["from"] = "delete";
            history.push("/redirect");
            deleteUser(user.email);
        } else {
            this.setState({
                deleteCodeIncorrect: true,
            });
            setTimeout(() => {
                this.setState({
                    deleteCodeIncorrect: false,
                });
            }, 3000);
        }
    }

    render() {
        const {
            displayEmailWindow,
            displayDeleteWindow,
            deleteCode,
            deleteCodeIncorrect,
            visible,
            authenticated,
            failed,
            loading,
        } = this.state;

        return (
            <main className='profile-container'>
                <MobileNav />
                <SidePanel />
                <div className='user-info-container'>
                    <section className='account'>
                        <h2 className='profile-heading' id='account-h2'>
                            My Account
                        </h2>
                        <AccountInfo
                            user={this.props.user}
                            displayEmailWindow={displayEmailWindow}
                            onClick={this.onClick}
                            cancelEdit={this.cancelEdit}
                            onChange={this.onChange}
                            onSubmit={this.onSubmit}
                            emailField={this.emailField}
                        />
                    </section>
                    <section className='password'>
                        <h2 className='profile-heading' id='password-h2'>
                            Password Settings
                        </h2>
                        <Password
                            visible={visible}
                            authenticated={authenticated}
                            displayPasswordChange={this.displayPasswordChange}
                            hidePassword={this.hidePassword}
                            onChange={this.onChange}
                            sendChangeCode={this.sendChangeCode}
                            submitCode={this.submitCode}
                            failed={failed}
                            verifyFailed={this.verifyFailed}
                            loading={loading}
                        />
                    </section>
                    <section className='privacy'>
                        <h2 className='profile-heading' id='privacy-h2'>
                            Privacy
                        </h2>
                        <section className='profile-form-container'>
                            <div className='profile-box'>
                                <p className='profile-label'>
                                    Profile visibility:
                                </p>
                                <div className='profile-visibility'>
                                    <input
                                        type='checkbox'
                                        id='profile-public'
                                        name='profile-public'
                                        defaultChecked
                                    />
                                    <label htmlFor='profile-public'>
                                        Public
                                    </label>
                                    <br />
                                    <input
                                        type='checkbox'
                                        id='profile-private'
                                        name='profile-private'
                                    />
                                    <label htmlFor='profile-private'>
                                        Private
                                    </label>
                                </div>
                            </div>
                            <div className='profile-box'>
                                <p className='profile-label'>
                                    Receive emails from us?
                                </p>
                                <div className='profile-visibility'>
                                    <input
                                        type='checkbox'
                                        id='receive-email-pos'
                                        name='receive-email-pos'
                                        defaultChecked
                                    />
                                    <label htmlFor='receive-email-pos'>
                                        Yes
                                    </label>
                                    <br />
                                    <input
                                        type='checkbox'
                                        id='receive-email-neg'
                                        name='receive-email-neg'
                                    />
                                    <label htmlFor='receive-email-neg'>
                                        No
                                    </label>
                                </div>
                            </div>
                        </section>
                    </section>
                    <section className='delete-account-container'>
                        <DeleteAccount
                            onClick={this.onClick}
                            onChange={this.onChange}
                            cancelEdit={this.cancelEdit}
                            displayDeleteWindow={displayDeleteWindow}
                            deleteField={this.deleteField}
                            onDelete={this.onDelete}
                            deleteCode={deleteCode}
                            deleteCodeIncorrect={deleteCodeIncorrect}
                        />
                    </section>
                </div>
            </main>
        );
    }
}

Account.propTypes = {
    user: PropTypes.object,
    displayAlert: PropTypes.func,
    updateUserEmail: PropTypes.func.isRequired,
    loadUser: PropTypes.func,
    deleteUser: PropTypes.func,
};

const setProps = (state) => ({
    user: state.auth.user,
});

export default connect(setProps, {
    displayAlert,
    updateUserEmail,
    deleteUser,
    loadUser,
})(withRouter(Account));
