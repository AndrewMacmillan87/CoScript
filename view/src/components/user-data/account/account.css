.profile-container {
    margin-left: 300px;
}

.profile-box {
    width: 70%;
    margin: 0 auto 20px auto !important;
}

.input-container {
    display: -webkit-box;
    display: flex;
}

.profile-label {
    display: block;
    font-weight: bold;
    margin-bottom: 5px;
}

.account-box {
    display: block;
    color: #111;
    border-top: none;
    border-right: none;
    border-bottom: 4px solid #444;
    border-left: none;
    width: 100%;
    font-size: 1rem;
    padding-left: 10px;
}

.users-email {
    display: block;
    padding-top: 8px;
}

.edit-icon-btn {
    background: none;
    border: none;
    height: 40px;
    margin-left: 15px;
    cursor: pointer;
}

.edit-icon-btn:focus {
    outline: 0;
}

.edit-input-icon {
    height: auto;
    width: 20px;
}

.edit-input-icon:hover {
    -webkit-transform: scale(1.2);
    transform: scale(1.2);
    -webkit-transition: 0.2s, -webkit-transform;
    transition: 0.2s, -webkit-transform;
    transition: transform, 0.2s;
    transition: transform, 0.2s, -webkit-transform;
}

.save-btn-container {
    text-align: center;
    margin: 0 auto;
}

#password-h2 {
    display: block;
    margin-bottom: 0 !important;
}

.pw-button-container {
    text-align: center;
    margin: 0 auto 15px auto;
}

.pw-hide-btn {
    background: none;
    border: none;
    padding: 0;
    cursor: pointer;
}

.pw-hide-btn:hover {
    color: var(--button-green, #4af030);
}

.pw-hide-btn:focus {
    outline: 0;
}

.prompt-text {
    display: block;
    padding-top: 15px;
}

.pw-code-failed {
    display: block;
    margin-top: 10px;
    color: red;
}

.pw-code-container {
    display: -webkit-box;
    display: flex;
    margin: 10px 0;
}

#new-pw,
#new-pw-confirm {
    padding-bottom: 10px;
}

.pw-code {
    display: -webkit-box;
    display: flex;
}

.pw-code-input {
    padding-left: 10px;
}

.pw-code-input:focus {
    outline: 0;
}

.verify-pw-code-spinner {
    margin-left: 20px;
    height: 40px;
    width: 40px;
}

#profile-public,
#receive-email-pos {
    margin-bottom: 15px;
}

.delete-account-container {
    border: 2px solid red;
    width: 70%;
    margin: 20px auto 50px auto;
    padding: 0 15px 15px 15px;
}

.delete-heading {
    display: block;
    color: red;
    margin-top: 10px;
}

.delete-disclaimer {
    display: block;
    font-size: 0.9rem;
    text-align: justify;
    padding: 5px 0;
}

.account-edit-container {
    position: fixed;
    margin: 0 auto;
    padding-right: 10px;
    top: 30%;
    right: 0;
    left: 0;
    width: 50%;
    background: rgb(235, 235, 235);
    box-shadow: 5px 5px 10px rgb(177, 177, 177);
}

.email-edit-input {
    width: 100%;
    height: 30px;
    padding-left: 10px;
}

.profile-edit-heading {
    display: block;
    padding: 20px 0 0 30px;
    font-weight: bold;
    font-size: 1.8rem;
}

.profile-edit-msg {
    display: block;
    padding: 20px 0 0 30px;
    font-size: 1rem;
}

.delete-error {
    color: red;
    padding: 0 0 10px 30px;
    font-size: 1rem;
}

@media (max-width: 500px) {
    .profile-container {
        margin-left: 0;
    }

    .side-panel-container {
        display: none;
    }

    .profile-visibility {
        margin: 15px 0 15px 5px;
    }

    .account-edit-container {
        width: 90%;
    }
}
