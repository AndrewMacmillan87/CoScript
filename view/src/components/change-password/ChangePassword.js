import React, { Component } from "react";
import { withRouter } from "react-router-dom";
import Alert from "../alert/Alert";
import PostButton from "../buttons/PostButton";

// Redux
import { connect } from "react-redux";
import { displayAlert } from "../../redux/actions/alertpopup";
import {
    checkCurrentPassword,
    comparePW,
} from "../../redux/auxiliary/password";
import { updatePassword } from "../../redux/actions/account";

import PropTypes from "prop-types";

import "./change-password.css";

export class ChangePassword extends Component {
    constructor(props) {
        super(props);
        this.state = {
            currentPassword: "",
            newPassword: "",
            confirmPassword: "",
            loading: false,
        };
        this.displayLoading = this.displayLoading.bind(this);
        this.removeLoading = this.removeLoading.bind(this);
    }

    /**
     * Sets the browser tab title and scrolls to
     * the top of the page when the component mounts
     */
    componentDidMount() {
        document.title = "CoScript - Change Password";
        window.scrollTo(0, 0);
    }

    /**
     * Updates the <[e.target.name]> state variable
     * with the text value of the event target
     *
     * @param {Object} e - event object
     */
    onChange(e) {
        this.setState({
            [e.target.name]: e.target.value,
        });
    }

    /**
     * Sets the loading state property to true to
     * display a loading spinner to the user while
     * processing
     */
    displayLoading() {
        this.setState({
            loading: true,
        });
    }

    /**
     * Sets the loading state property to false to
     * remove a loading spinner once processing has
     * completed
     */
    removeLoading() {
        this.setState({
            loading: false,
        });
    }

    /**
     * Performs error checking on the form for:
     *  - current password being correct
     *  - the new password clashing with a previously used password
     *  - new password and new password confirmation not matching
     *  - blank form fields
     *
     * If no errors are found, the user's new password is
     * sent to the server to be updated in the database
     *
     * @param {Object} e - event object
     */
    async onSubmit(e) {
        e.preventDefault();
        this.displayLoading();
        const { currentPassword, newPassword, confirmPassword } = this.state;
        const { user, displayAlert, updatePassword } = this.props;
        const checkCurrentPW = await checkCurrentPassword(
            user,
            currentPassword
        );
        const equal = newPassword === confirmPassword;
        const passwordClash = await comparePW(user, newPassword);
        const blank = !newPassword && !confirmPassword;
        if (!checkCurrentPW) {
            displayAlert("Your current password is not correct", "danger");
            this.removeLoading();
            return;
        } else if (!equal) {
            displayAlert("Passwords do not match", "danger");
            this.removeLoading();
            return;
        } else if (passwordClash) {
            displayAlert("That's an old password", "danger");
            this.removeLoading();
            return;
        } else if (blank) {
            displayAlert("Fields cannot be left blank", "danger");
            this.removeLoading();
            return;
        }
        await updatePassword(newPassword);
    }

    render() {
        const { loading } = this.state;
        const spinner = require("../../img/loading.svg");

        return (
            <main className='change-password-main'>
                <section className='change-password-section'>
                    <h1 className='change-password-heading'>CoScript</h1>
                    <h2 className='change-password-subheading'>
                        Update your password
                    </h2>
                    <div className='password-box'>
                        <form
                            className='new-pw-form'
                            onSubmit={(e) => this.onSubmit(e)}>
                            <div className='input-account-container'>
                                <label
                                    className='profile-label'
                                    htmlFor='new-pw'>
                                    Current Password
                                </label>
                                <input
                                    type='password'
                                    name='currentPassword'
                                    className='account-input'
                                    onChange={(e) => this.onChange(e)}
                                    placeholder='Current password...'
                                    onFocus={(e) => (e.target.placeholder = "")}
                                    onBlur={(e) =>
                                        (e.target.placeholder =
                                            "Current password...")
                                    }
                                />
                            </div>
                            <div className='input-account-container'>
                                <label
                                    className='profile-label'
                                    htmlFor='new-pw'>
                                    New Password
                                </label>
                                <input
                                    type='password'
                                    name='newPassword'
                                    className='account-input'
                                    onChange={(e) => this.onChange(e)}
                                    placeholder='New password...'
                                    onFocus={(e) => (e.target.placeholder = "")}
                                    onBlur={(e) =>
                                        (e.target.placeholder =
                                            "New password...")
                                    }
                                />
                            </div>
                            <div className='input-account-container'>
                                <label
                                    className='profile-label'
                                    htmlFor='new-pw-confirm'>
                                    Confirm Password
                                </label>
                                <input
                                    type='password'
                                    name='confirmPassword'
                                    className='account-input'
                                    onChange={(e) => this.onChange(e)}
                                    placeholder='Confirm password...'
                                    onFocus={(e) => (e.target.placeholder = "")}
                                    onBlur={(e) =>
                                        (e.target.placeholder =
                                            "Confirm password...")
                                    }
                                />
                            </div>
                            <Alert />
                            <div className='pw-change-btn-container'>
                                <PostButton
                                    classes='small-btn'
                                    value='Update'
                                />
                                {loading && (
                                    <img
                                        className='password-spinner'
                                        src={spinner}
                                        alt='loading'
                                    />
                                )}
                            </div>
                        </form>
                    </div>
                </section>
            </main>
        );
    }
}

ChangePassword.propTypes = {
    user: PropTypes.object,
    updatePassword: PropTypes.func.isRequired,
    displayAlert: PropTypes.func.isRequired,
};

const setProps = (state) => ({
    user: state.auth.user,
});

export default connect(setProps, {
    updatePassword,
    displayAlert,
})(withRouter(ChangePassword));
