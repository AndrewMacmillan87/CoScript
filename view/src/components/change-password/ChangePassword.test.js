import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";

import { Provider } from "react-redux";
import { reduxStore } from "../../redux/store/reduxStore";

import { Alert } from "../alert/Alert";
import { ChangePassword } from "./ChangePassword";
import PostButton from "../buttons/PostButton";

import * as pwFuncs from "../../redux/auxiliary/password";
import * as accountActions from "../../redux/actions/account";

const originalCheck = pwFuncs.checkCurrentPassword;
const originalCompare = pwFuncs.comparePW;

describe("ChangePassword component", () => {
    let mockUpdatePassword, mockDisplayAlert, props;
    beforeEach(() => {
        global.scrollTo = jest.fn();
        mockUpdatePassword = jest.fn();
        mockDisplayAlert = jest.fn();
        props = {
            updatePassword: mockUpdatePassword,
            displayAlert: mockDisplayAlert,
        };
    });

    afterEach(() => {
        pwFuncs.checkCurrentPassword = originalCheck;
        pwFuncs.comparePW = originalCompare;
    });

    test("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <ChangePassword {...props} />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a main element with a className of change-password-main", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("main").length).toEqual(1);
        expect(
            wrapper.find({ className: "change-password-main" }).length
        ).toEqual(1);
    });

    it("should render a section element with a className of change-password-section", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("section").length).toEqual(1);
        expect(
            wrapper.find({ className: "change-password-section" }).length
        ).toEqual(1);
    });

    it("should render a form element with a className of new-pw-form", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("form").length).toEqual(1);
        expect(wrapper.find({ className: "new-pw-form" }).length).toEqual(1);
    });

    it("should render 5 div elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("div").length).toEqual(5);
    });

    it("should render 3 label elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("label").length).toEqual(3);
    });

    it("should render 4 input elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("input").length).toEqual(4);
    });

    it("should render a PostButton component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(PostButton).length).toEqual(1);
    });

    it("should render an Alert component if checkCurrentPassword is false", async () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );

        pwFuncs.checkCurrentPassword = jest.fn(() => false);

        const fakeEvent = {
            preventDefault: () => {},
        };

        wrapper
            .find("input")
            .at(0)
            .simulate("change", {
                target: { name: "currentPassword", value: null },
            });
        wrapper
            .find("input")
            .at(1)
            .simulate("change", {
                target: { name: "newPassword", value: "password4321" },
            });
        wrapper
            .find("input")
            .at(2)
            .simulate("change", {
                target: { name: "confirmPassword", value: "password4321" },
            });

        await act(async () => {
            wrapper.find("form").simulate("submit", fakeEvent);
        });

        expect(pwFuncs.checkCurrentPassword).toBeCalled();
        expect(wrapper.find(Alert).length).toEqual(1);
    });

    it("should render an Alert component if password and confirmPassword do not match", async () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );

        pwFuncs.checkCurrentPassword = jest.fn(() => true);
        pwFuncs.comparePW = jest.fn(() => false);

        const fakeEvent = {
            preventDefault: () => {},
        };

        wrapper
            .find("input")
            .at(0)
            .simulate("change", {
                target: { name: "currentPassword", value: "password1234" },
            });
        wrapper
            .find("input")
            .at(1)
            .simulate("change", {
                target: { name: "newPassword", value: "1234drowssap" },
            });
        wrapper
            .find("input")
            .at(2)
            .simulate("change", {
                target: { name: "confirmPassword", value: "password4321" },
            });

        await act(async () => {
            wrapper.find("form").simulate("submit", fakeEvent);
        });

        expect(pwFuncs.checkCurrentPassword).toBeCalled();
        expect(pwFuncs.comparePW).toBeCalled();
        expect(wrapper.find(Alert).length).toEqual(1);
    });

    it("should render an Alert component if passwordClash is true", async () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );

        pwFuncs.checkCurrentPassword = jest.fn(() => true);
        pwFuncs.comparePW = jest.fn(() => true);

        const fakeEvent = {
            preventDefault: () => {},
        };

        wrapper
            .find("input")
            .at(0)
            .simulate("change", {
                target: { name: "currentPassword", value: "password1234" },
            });
        wrapper
            .find("input")
            .at(1)
            .simulate("change", {
                target: { name: "newPassword", value: "password4321" },
            });
        wrapper
            .find("input")
            .at(2)
            .simulate("change", {
                target: { name: "confirmPassword", value: "password4321" },
            });

        await act(async () => {
            wrapper.find("form").simulate("submit", fakeEvent);
        });

        expect(pwFuncs.checkCurrentPassword).toBeCalled();
        expect(pwFuncs.comparePW).toBeCalled();
        expect(wrapper.find(Alert).length).toEqual(1);
    });

    it("should render an Alert component if a new password field is blank", async () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );

        pwFuncs.checkCurrentPassword = jest.fn(() => true);
        pwFuncs.comparePW = jest.fn(() => true);

        const fakeEvent = {
            preventDefault: () => {},
        };

        wrapper
            .find("input")
            .at(0)
            .simulate("change", {
                target: { name: "currentPassword", value: "password1234" },
            });
        wrapper
            .find("input")
            .at(1)
            .simulate("change", {
                target: { name: "newPassword", value: "" },
            });
        wrapper
            .find("input")
            .at(2)
            .simulate("change", {
                target: { name: "confirmPassword", value: "password4321" },
            });

        await act(async () => {
            wrapper.find("form").simulate("submit", fakeEvent);
        });

        expect(pwFuncs.checkCurrentPassword).toBeCalled();
        expect(pwFuncs.comparePW).toBeCalled();
        expect(wrapper.find(Alert).length).toEqual(1);
    });

    test("should call `checkCurrentPassword()`, `comparePW()` and `updatePassword()` functions when form is submitted with correct data", async () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <ChangePassword {...props} />
                </Router>
            </Provider>
        );

        pwFuncs.checkCurrentPassword = jest.fn(() => true);
        pwFuncs.comparePW = jest.fn(() => false);
        accountActions.updatePassword = jest.fn(() => Promise.resolve(true));

        const fakeEvent = {
            preventDefault: () => {},
        };

        wrapper
            .find("input")
            .at(0)
            .simulate("change", {
                target: { name: "currentPassword", value: "password1234" },
            });
        wrapper
            .find("input")
            .at(1)
            .simulate("change", {
                target: { name: "newPassword", value: "password4321" },
            });
        wrapper
            .find("input")
            .at(2)
            .simulate("change", {
                target: { name: "confirmPassword", value: "password4321" },
            });

        await act(async () => {
            wrapper.find("form").simulate("submit", fakeEvent);
        });

        expect(pwFuncs.checkCurrentPassword).toBeCalled();
        expect(pwFuncs.comparePW).toBeCalled();
        expect(mockUpdatePassword).toBeCalled();
    });
});
