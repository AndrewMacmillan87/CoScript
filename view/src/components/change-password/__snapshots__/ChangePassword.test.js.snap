// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePassword component renders correctly 1`] = `
<main
  className="change-password-main"
>
  <section
    className="change-password-section"
  >
    <h1
      className="change-password-heading"
    >
      CoScript
    </h1>
    <h2
      className="change-password-subheading"
    >
      Update your password
    </h2>
    <div
      className="password-box"
    >
      <form
        className="new-pw-form"
        onSubmit={[Function]}
      >
        <div
          className="input-account-container"
        >
          <label
            className="profile-label"
            htmlFor="new-pw"
          >
            Current Password
          </label>
          <input
            className="account-input"
            name="currentPassword"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Current password..."
            type="password"
          />
        </div>
        <div
          className="input-account-container"
        >
          <label
            className="profile-label"
            htmlFor="new-pw"
          >
            New Password
          </label>
          <input
            className="account-input"
            name="newPassword"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="New password..."
            type="password"
          />
        </div>
        <div
          className="input-account-container"
        >
          <label
            className="profile-label"
            htmlFor="new-pw-confirm"
          >
            Confirm Password
          </label>
          <input
            className="account-input"
            name="confirmPassword"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Confirm password..."
            type="password"
          />
        </div>
        <div
          className="pw-change-btn-container"
        >
          <input
            className="small-btn"
            type="submit"
            value="Update"
          />
        </div>
      </form>
    </div>
  </section>
</main>
`;
